Archive/Imo/Imo2024Q1.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2024 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Algebra.BigOperators.Intervals
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Algebra.Order.ToIntervalMod
+import Mathlib.Data.Real.Archimedean
+import Mathlib.Tactic.Peel
+
+/-!
+# IMO 2024 Q1
+
+Determine all real numbers $\alpha$ such that, for every positive integer $n$, the integer
+\[
+\lfloor \alpha \rfloor + \lfloor 2\alpha \rfloor + \cdots + \lfloor n\alpha \rfloor
+\]
+is a multiple of~$n$.
+
+We follow Solution 3 from the
+[official solutions](https://www.imo2024.uk/s/IMO-2024-Paper-1-Solutions.pdf).  First reducing
+modulo 2, any answer that is not a multiple of 2 is inductively shown to be contained in a
+decreasing sequence of intervals, with empty intersection.
+-/
+
+
+namespace Imo2024Q1
+
+/-- The condition of the problem. -/
+def Condition (α : ℝ) : Prop := (∀ n : ℕ, 0 < n → (n : ℤ) ∣ ∑ i ∈ Finset.Icc 1 n, ⌊i * α⌋)
+
+lemma condition_two_mul_int (m : ℤ) : Condition (2 * m) := by
+  rintro n -
+  suffices (n : ℤ) ∣ ∑ i ∈ Finset.Icc 1 n, ⌊((i * (2 * m) : ℤ) : ℝ)⌋ from mod_cast this
+  suffices (n : ℤ) ∣ ∑ i ∈ Finset.Icc 0 n, ⌊((i * (2 * m) : ℤ) : ℝ)⌋ by
+    convert this using 1
+    convert (Finset.sum_insert_zero (a := 0) ?_).symm
+    · exact (Nat.Icc_insert_succ_left (Nat.zero_le n)).symm
+    · simp
+  simp_rw [Int.floor_intCast, ← Finset.sum_mul, ← Nat.Ico_succ_right, Nat.succ_eq_add_one,
+           ← Finset.range_eq_Ico, ← mul_assoc]
+  refine dvd_mul_of_dvd_left ?_ _
+  suffices (n : ℤ) ∣ ((n + 1) * (n + 1 - 1) : ℕ) by
+    convert this
+    rw [← Finset.sum_range_id_mul_two]
+    push_cast
+    rfl
+  simp
+
+lemma condition_sub_two_mul_int_iff {α : ℝ} (m : ℤ) : Condition (α - 2 * m) ↔ Condition α := by
+  simp_rw [Condition]
```

Comment:
`simp_rw` feels like overkill here. `unfold Condition` or `dsimp only [Condition]` would suffice.

Reply:
Changed to use `unfold`.